app/invocation/invocation_action_card.tsx

```diff
@@ -319,122 +319,161 @@ export default class InvocationActionCardComponent extends React.Component<Props
-                </span>
-              </div>
-              <div className="action-timeline">
-                <div
-                  className="timeline-event-gray"
-                  title={`${event.name} (${format.durationSec(duration)}, ${(weight * 100).toFixed(2)}%)`}
-                  style={{ flex: `${offset - weight} 0 0`, backgroundColor: `rgba(0, 0, 0, .1)` }}></div>
-                <div
-                  className="timeline-event"
-                  title={`${event.name} (${format.durationSec(duration)}, ${(weight * 100).toFixed(2)}%)`}
-                  style={{ flex: `${weight} 0 0`, backgroundColor: event.color }}></div>
-                <div
-                  className="timeline-event-gray"
-                  title={`${event.name} (${format.durationSec(duration)}, ${(weight * 100).toFixed(2)}%)`}
-                  style={{ flex: `${1 - offset} 0 0`, backgroundColor: `rgba(0, 0, 0, .1)` }}></div>
+      <>
+        <div className="metadata-title">Timeline</div>
+        <div>
+          <div className="metadata-detail">
+            <span className="label">Total @ {format.formatTimestamp(filteredEvents[0].timestamp)}</span>
+            <span className="bar-description">{format.durationSec(totalDuration)} (100%)</span>
+          </div>
+          <div className="action-timeline">
+            <div
+              className="timeline-event"
+              title={`Total: (${format.durationSec(totalDuration)}, 100%)`}
+              style={{ flex: `1 0 0`, backgroundColor: "green" }}></div>
+          </div>
+          {filteredEvents.map((event, i) => {
+            // Don't render the end marker.
+            if (i == filteredEvents.length - 1) return null;
+
+            const next = filteredEvents[i + 1];
+            if (timestampToUnixSeconds(event.timestamp) > timestampToUnixSeconds(next.timestamp)) {
+              // Queued event is recorded by the scheduler (not the workers)
+              // which may result in confusing timestamps due to clock drift.
+              // To reduce confusion, apply an adjustment here to enforce
+              // that all timestamps appear in monotonically increasing order
+              // when rendered as a timeline.
+              event.timestamp = next.timestamp;
+            }
+            const duration = durationSeconds(event.timestamp!, next.timestamp!);
+            const weight = duration / totalDuration;
+
+            // Actual flex initial value.
+            //
+            // Time drift between app and executor could cause negative duration,
+            // which is not a valid flex initial value and causes broken render.
+            // Make sure to enforce a minimum value of zero for "flex" style so
+            // it is always included in the final render result.
+            const leftBar = Math.max(0, offset);
+            const middleBar = Math.max(0, weight);
+            const rightBar = Math.max(0, 1 - leftBar - middleBar);
+
```

Comment:
I think all of these changes can be reverted now?

Reply:
Make sense to revert the top hunk here.

But I will leave the leftBar -> rightBar hunk in so that these are calculated before adding up the offset, which is a bit easier to read.